### PR TITLE
add  `hookActionValidateOrder` concrete use case example

### LINKED ISSUE #1539: Add `hooks`  implementations `example`
# Add `hooks`  implementation `examples`.

| Hook name | is documented |has example|
|:----------|:-------------:|------:|
| actionValidateOrder| ✅ | https://github.com/PrestaShop/docs/pull/1543 |
| actionValidateOrderAfter| ✅ | ❌ |
| displayMaintenance| ✅ | ❌ |
| displayCartModalContent| ✅ | ❌ |
| displayCartModalFooter| ✅ | ❌ |
| displayProductPageDrawer| ✅ | ❌ |
| actionPaymentConfirmation| ✅ | ❌ |
| displayPaymentReturn| ✅ | ❌ |
| actionUpdateQuantity| ✅ | ❌ |
| displayRightColumn| ✅ | ❌ |
| displayWrapperTop| ✅ | ❌ |
| displayWrapperBottom| ✅ | ❌ |
| displayContentWrapperTop| ✅ | ❌ |
| displayContentWrapperBottom| ✅ | ❌ |
| displayLeftColumn| ✅ | ❌ |
| displayHome| ✅ | ❌ |
| displayHeader| ✅ | ❌ |
| actionCartSave| ✅ | ❌ |
| actionAuthentication| ✅ | ❌ |
| actionProductAdd| ✅ | ❌ |
| actionProductUpdate| ✅ | ❌ |
| displayAfterTitleTag| ✅ | ❌ |
| displayAfterBodyOpeningTag| ✅ | ❌ |
| displayBanner| ✅ | ❌ |
| displayBeforeBodyClosingTag| ✅ | ❌ |
| displayTop| ✅ | ❌ |
| displayNavFullWidth| ✅ | ❌ |
| displayRightColumnProduct| ✅ | ❌ |
| actionProductDelete| ✅ | ❌ |
| actionObjectProductInCartDeleteBefore| ✅ | ❌ |
| actionObjectProductInCartDeleteAfter| ✅ | ❌ |
| displayFooterProduct| ✅ | ❌ |
| displayInvoice| ✅ | ❌ |
| actionOrderStatusUpdate| ✅ | ❌ |
| displayAdminGridTableBefore| ✅ | ❌ |
| displayAdminGridTableAfter| ✅ | ❌ |
| displayAdminOrder| ✅ | ❌ |
| displayAdminOrderTabOrder| ✅ | ❌ |
| displayAdminOrderTabShip| ✅ | ❌ |
| displayAdminOrderContentOrder| ✅ | ❌ |
| displayAdminOrderContentShip| ✅ | ❌ |
| displayFooter| ✅ | ❌ |
| displayInvoiceLegalFreeText| ✅ | ❌ |
| displayAdminCustomers| ✅ | ❌ |
| displayAdminCustomersAddressesItemAction| ✅ | ❌ |
| displayOrderConfirmation| ✅ | ❌ |
| actionCustomerAccountAdd| ✅ | ❌ |
| actionCustomerAccountUpdate| ✅ | ❌ |
| displayCustomerAccount| ✅ | ❌ |
| actionOrderSlipAdd| ✅ | ❌ |
| displayShoppingCartFooter| ✅ | ❌ |
| displayCustomerAccountForm| ✅ | ❌ |
| displayAdminStatsModules| ✅ | ❌ |
| actionOrderReturn| ✅ | ❌ |
| displayProductAdditionalInfo| ✅ | ❌ |
| actionWatermark| ✅ | ❌ |
| actionProductCancel| ✅ | ❌ |
| displayLeftColumnProduct| ✅ | ❌ |
| actionProductOutOfStock| ✅ | ❌ |
| actionProductAttributeUpdate| ✅ | ❌ |
| displayCarrierList| ✅ | ❌ |
| displayShoppingCart| ✅ | ❌ |
| actionCarrierUpdate| ✅ | ❌ |
| actionOrderStatusPostUpdate| ✅ | ❌ |
| displayCustomerAccountFormTop| ✅ | ❌ |
| displayBackOfficeHeader| ✅ | ❌ |
| displayBackOfficeTop| ✅ | ❌ |
| displayAdminEndContent| ✅ | ❌ |
| displayBackOfficeFooter| ✅ | ❌ |
| actionProductAttributeDelete| ✅ | ❌ |
| actionCarrierProcess| ✅ | ❌ |
| displayBeforeCarrier| ✅ | ❌ |
| displayAfterCarrier| ✅ | ❌ |
| displayOrderDetail| ✅ | ❌ |
| actionPaymentCCAdd| ✅ | ❌ |
| actionCategoryAdd| ✅ | ❌ |
| actionCategoryUpdate| ✅ | ❌ |
| actionCategoryDelete| ✅ | ❌ |
| displayPaymentTop| ✅ | ❌ |
| actionHtaccessCreate| ✅ | ❌ |
| actionAdminMetaSave| ✅ | ❌ |
| displayAttributeGroupForm| ✅ | ❌ |
| actionAttributeGroupSave| ✅ | ❌ |
| actionAttributeGroupDelete| ✅ | ❌ |
| displayFeatureForm| ✅ | ❌ |
| actionFeatureSave| ✅ | ❌ |
| actionFeatureDelete| ✅ | ❌ |
| actionProductSave| ✅ | ❌ |
| displayFeaturePostProcess| ✅ | ❌ |
| displayFeatureValueForm| ✅ | ❌ |
| displayFeatureValuePostProcess| ✅ | ❌ |
| actionFeatureValueDelete| ✅ | ❌ |
| actionFeatureValueSave| ✅ | ❌ |
| displayAttributeForm| ✅ | ❌ |
| actionAttributeDelete| ✅ | ❌ |
| actionAttributeSave| ✅ | ❌ |
| displayMyAccountBlock| ✅ | ❌ |
| actionModuleInstallBefore| ✅ | ❌ |
| actionModuleInstallAfter| ✅ | ❌ |
| displayBackOfficeCategory| ✅ | ❌ |
| displayNav| ✅ | ❌ |
| actionAdminLoginControllerSetMedia| ✅ | ❌ |
| actionOrderEdited| ✅ | ❌ |
| actionEmailAddBeforeContent| ✅ | ❌ |
| actionEmailAddAfterContent| ✅ | ❌ |
| sendMailAlterTemplateVars| ✅ | ❌ |
| displayCartExtraProductActions| ✅ | ❌ |
| displayPaymentByBinaries| ✅ | ❌ |
| additionalCustomerFormFields| ✅ | ❌ |
| additionalCustomerAddressFields| ✅ | ❌ |
| addWebserviceResources| ✅ | ❌ |
| displayCustomerLoginFormAfter| ✅ | ❌ |
| actionClearCache| ✅ | ❌ |
| actionClearCompileCache| ✅ | ❌ |
| actionClearSf2Cache| ✅ | ❌ |
| actionValidateCustomerAddressForm| ✅ | ❌ |
| displayCarrierExtraContent| ✅ | ❌ |
| validateCustomerFormFields| ✅ | ❌ |
| displayProductExtraContent| ✅ | ❌ |
| filterCmsContent| ✅ | ❌ |
| filterCmsCategoryContent| ✅ | ❌ |
| filterProductContent| ✅ | ❌ |
| filterCategoryContent| ✅ | ❌ |
| filterManufacturerContent| ✅ | ❌ |
| filterSupplierContent| ✅ | ❌ |
| filterHtmlContent| ✅ | ❌ |
| displayDashboardTop| ✅ | ❌ |
| actionUpdateLangAfter| ✅ | ❌ |
| actionOutputHTMLBefore| ✅ | ❌ |
| displayAfterProductThumbs| ✅ | ❌ |
| actionDispatcherBefore| ✅ | ❌ |
| actionDispatcherAfter| ✅ | ❌ |
| filterProductSearch| ✅ | ❌ |
| actionProductSearchAfter| ✅ | ❌ |
| actionEmailSendBefore| ✅ | ❌ |
| displayAdminProductsMainStepLeftColumnMiddle| ✅ | ❌ |
| displayAdminProductsMainStepLeftColumnBottom| ✅ | ❌ |
| displayAdminProductsMainStepRightColumnBottom| ✅ | ❌ |
| displayAdminProductsQuantitiesStepBottom| ✅ | ❌ |
| displayAdminProductsPriceStepBottom| ✅ | ❌ |
| displayAdminProductsOptionsStepTop| ✅ | ❌ |
| displayAdminProductsOptionsStepBottom| ✅ | ❌ |
| displayAdminProductsSeoStepBottom| ✅ | ❌ |
| displayAdminProductsShippingStepBottom| ✅ | ❌ |
| displayAdminProductsExtra| ✅ | ❌ |
| displayAdminProductsCombinationBottom| ✅ | ❌ |
| displayDashboardToolbarTopMenu| ✅ | ❌ |
| displayDashboardToolbarIcons| ✅ | ❌ |
| actionFrontControllerInitAfter| ✅ | ❌ |
| actionFrontControllerInitBefore| ✅ | ❌ |
| actionProductFlagsModifier| ✅ | ❌ |
| displayProductActions| ✅ | ❌ |
| displayPersonalInformationTop| ✅ | ❌ |
| displayAdminOrderTop| ✅ | ❌ |
| displayBackOfficeOrderActions| ✅ | ❌ |
| displayAdminOrderSide| ✅ | ❌ |
| displayAdminOrderMain| ✅ | ❌ |
| displayAdminOrderMainBottom| ✅ | ❌ |
| displayAdminOrderTabLink| ✅ | ❌ |
| displayAdminOrderTabContent| ✅ | ❌ |
| actionGetAdminOrderButtons| ✅ | ❌ |
| actionAdminAdminPreferencesControllerPostProcessBefore| ✅ | ❌ |
| displayAdditionalCustomerAddressFields| ✅ | ❌ |
| displayFooterCategory| ✅ | ❌ |
| displayHeaderCategory| ✅ | ❌ |
| actionFrontControllerSetVariables| ✅ | ❌ |
| displayAdminOrderCreateExtraButtons| ✅ | ❌ |
| displayAdminGridTableBefore| ✅ | ❌ |
| displayAdminGridTableAfter| ✅ | ❌ |
| actionAdminOrdersTrackingNumberUpdate| ✅ | ❌ |
| displayBackOfficeEmployeeMenu| ✅ | ❌ |
| actionCustomerLogoutBefore| ✅ | ❌ |
| actionCustomerLogoutAfter| ✅ | ❌ |
| displayCheckoutSummaryTop| ✅ | ❌ |
| actionPresentPaymentOptions| ✅ | ❌ |
| dashboardZoneThree| ✅ | ❌ |
| actionFilterDeliveryOptionList| ✅ | ❌ |
| displayAdminThemesListAfter| ✅ | ❌ |
| actionGetAdminToolbarButtons| ✅ | ❌ |
| displayEmptyModuleCategoryExtraMessage| ✅ | ❌ |
| actionGetAlternativeSearchPanels| ✅ | ❌ |

### PR DESCRIPTION
**Pull request by PululuK**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 8.x 
| Description?  | add  `hookActionValidateOrder` concrete use case example
| Fixed ticket? | Fixes #1539

### CODE REVIEW

File: modules/concepts/hooks/list-of-hooks.md [hunk 1]
```diff
@@ -1154,6 +1154,78 @@ actionValidateOrder
...
+                   $hasCustomerRequiredNbrOfTheOrderToReward = $this->getCustomerValidOrdersNbr((int) $customerObject->id) == $this->getRequiredNbrOfTheOrderToReward();
+
+                   if($hasConfiguredState && $hasCustomerRequiredNbrOfTheOrderToReward) {
+                       $customerReward = $this->createCustomerReward($customerObject, $orderObject);
+
+                       if(Validate::isLoadedObject($customerReward)) {
+                             $this->setAlreadyRewarded($customerObject);
+                             $this->notifyCustomer($customerObject, $customerReward);
+			     
+			     //TODO : of course don't forget to log if something fail here :)
```
Comment (thomasnares) [inline modules/concepts/hooks/list-of-hooks.md]:
```suggestion
			     //TODO : of course don't forget to log if something fails here :)
```

File: modules/concepts/hooks/list-of-hooks.md [hunk 2]
```diff
@@ -1154,6 +1154,78 @@ actionValidateOrder
...
+            }
+
+            protedect function getCustomerValidOrdersNbr(int $idCustomer): int
+            {
+               //TODO : return number of total order valid by customer
+            }
+
+            protected function getRequiredNbrOfTheOrderToReward(): int
+            {
+               //TODO : return configured requored number of the order to reward the customer
```
Comment (thomasnares) [inline modules/concepts/hooks/list-of-hooks.md]:
```suggestion
               //TODO : return configured number of orders required to reward the customer
```

File: modules/concepts/hooks/list-of-hooks.md [hunk 3]
```diff
@@ -1154,6 +1154,78 @@ actionValidateOrder
...
+            }
+
+            protected function createCustomerReward(Customer $customer, Order $order): ?CartRule
+            {
+               //TODO: generate customer cart rule (according the order amount for example)
```
Comment (thomasnares) [inline modules/concepts/hooks/list-of-hooks.md]:
```suggestion
               //TODO: generate customer cart rule (according to the order amount for example)
```

**Review by thomasnares — COMMENTED** [review]
Hello @PululuK , just some suggestions of wording corrections. Thanks 👍

**Comment by kpodemski** [comment]
Hello @PululuK 

My opinion is that having examples on this page is too much. We could think of a more advanced module showing the implementation of many hooks, to which we could later link.

It is probably something that  @thomasnares and @eternoendless should decide on.

I agree that having more real-life examples could be beneficial for developers. I was thinking of some _recepies_ somewhere in the docs. Another option is to continue working on the example-modules repository.

**Comment by kpodemski** [comment]
Hi @PululuK 

As you know, @thomasnares improved documentation to have such examples on separate pages. This is yours:
https://devdocs.prestashop-project.org/8/modules/sample-modules/example-hooks/actionvalidateorder/

I hope this can unblock your challenge 😎

**Comment by PululuK** [comment]
Thanks @kpodemski @thomasnares !!!

**Pull request closed without merging**